### DISCUSSION TITLE
Is it possible to compress all files?

### DISCUSSION THREAD

**Discussion by AndreyRGW** [upx/upx#663] [Q&A] [answered]
Hello, everyone! is it possible to compress all exe and dll files in a folder with UPX? 

I am using Windows 11.

**Comment by jreiser** [comment, marked as answer]
Running the command `upx --help` shows:

                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2023
     UPX git-XXXXXXX Markus Oberhumer, Laszlo Molnar & John Reiser    Feb Xth 2023
     
     Usage: upx [-123456789dlthVL] [-qvfk] [-o file] file..

The notation `file...` with trailing ellipsis (the three dots) means that a list of multiple filenames may appear on the command line, separated by spaces or tabs (and not by commas.) How to express that list is a matter between you and your shell program.  On UNIX, Linux, and MacOS all shells expand metacharacters including `*?` (asterisk and question mark) to a list of actual filenames which match the pattern.  On Windows a traditional CMD shell does not do this, so you must find some other way to put the list into a command line.  In PowerShell probably there is an option.
Then there is a limit on the number of characters in a command line, which is up to your shell and the OS.  It may be 32K, 64K, 128K, or something larger.  UPX also has a size limit on any input file, something a little less than 900MiB.  Particularly on Windows, various software build systems might construct .exe and/or .dll which do not have the standard layout; UPX might decline to process such "exotic" files.

**Reply by AndreyRGW** [reply]
> How to express that list is a matter between you and your shell program. On UNIX, Linux, and MacOS all shells expand metacharacters including *? (asterisk and question mark) to a list of actual filenames which match the pattern. On Windows a traditional CMD shell does not do this, so you must find some other way to put the list into a command line.

On Windows 11, the new terminal accepts "*?". Many thanks!

**Reply by AndreyRGW** [reply]
By the way, do you know how to apply UPX to all subdirectories?

**Reply by jreiser** [reply]
UPX has no builtin which implements "traverse all subdirectories".  Use your shell.   On *nix:    `upx  $(find .  - name '*.exe' -o -name '*.dll')`
